relay-kafka/src/limits.rs

```diff
@@ -0,0 +1,103 @@
+use std::cmp;
+use std::collections::HashMap;
+use std::fmt;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::time::{Duration, Instant};
+
+use parking_lot::RwLock;
+
+use crate::debounced::Debounced;
+
+type Key = [u8; 16];
+
+/// A really basic rate limiter to protect downstream systems from partition imbalance on ingestion
+/// topics.
+///
+/// The rate limiter is not used to drop messages, but rather to have them be randomly partitioned
+/// (instead of by trace ID) to reduce the partition imbalance again. If the topic is semantically
+/// partitioned this almost certainly will have product impact.
+///
+/// A partition key's message rate is measured in a single window. If the span rate ever exceeds this rate
+/// (`limit_per_window / window_size`), the partition key is limited for up to a duration of
+/// `window_size`.
+#[derive(Clone)]
+pub struct KafkaRateLimits {
+    limit_per_window: u64,
+    map: Arc<RwLock<HashMap<Key, AtomicU64>>>,
+    wipe_debouncer: Arc<Debounced>,
+    window_size: Duration,
+}
+
+impl KafkaRateLimits {
+    pub fn new(limit_per_window: u64, window_size: Duration) -> Self {
+        KafkaRateLimits {
+            map: Default::default(),
+            wipe_debouncer: Arc::new(Debounced::new(window_size.as_secs())),
+            limit_per_window,
+            window_size,
+        }
+    }
+    #[must_use]
+    pub fn try_increment(&self, now: Instant, key: &Key, amount: u64) -> u64 {
+        self.wipe_debouncer.debounce(now, || {
+            self.map.write().clear();
+        });
+
+        let increment = |value: &AtomicU64| {
+            // we're ok with a few data races, we just want to be fast and avoid locks.
+            // fetch_update is probably slower since it has stronger atomicity guarantees.
+            let granted = cmp::min(amount, value.load(Ordering::Relaxed));
+            value.fetch_sub(granted, Ordering::Relaxed);
+            granted
```

Comment:
I think it's better if you count up and compare against the limit, this will also make this not racy, as this is a race condition between loading and subtracting, which will underflow (which is worse than just making one wrong decision).

When you count up you can always unconditionally increment and the returned value from `fetch_add` you can compare against the limit.

Reply:
ah nice. yeah the overflow is actually tolerable